db/src/db.ts

```diff
@@ -85,17 +147,55 @@ export class DB {
    * @return - The new value returned from updater
    */
   public async update<T extends Serializable = any>(
-    key: string, updater: (state?: T) => T
+    key: string, updater: (state?: T) => T, options?: Partial<UpdateOptions>,
   ): Promise<T> {
     return await this.writeLock.runExclusive(async () => {
-      const prev = await this.get(key);
-      const next = updater(prev as T);
-      checkValue(next);
-      await this.db.put(key, JSON.stringify(next));
-      return next;
+      let prev: StoredDocument<T> | Tombstone | undefined = await this.getWithMeta<T>(key);
+      if (prev === undefined) {
+        prev = { version: 0, value: undefined, patches: [], updatedAt: 0 };
+      }
+      const { version: prevVersion, value: prevValue, patches } = prev;
+      deepFreeze(prev);
+      const nextValue = updater(prevValue);
+      checkValue(nextValue);
+      const nextVersion = prevVersion + 1;
+      const ops = compare({ root: prevValue }, { root: nextValue });
+      if (ops.length > 0) {
+        const patch = { version: nextVersion, ops, ...options };
+        await this.db.put(key, JSON.stringify({
+          version: nextVersion,
+          value: nextValue,
+          patches: patches.slice(-NUM_PATCHES_TO_KEEP).concat(patch),
+          updatedAt: Date.now(),
+        }));
+        this.emit('patch', key, patch);
+      }
+      return nextValue;
     });
   }
 
+  public async poll(keysToVersions: KeyedVersions, opts: Partial<PollOptions> = {}): Promise<KeyedPatches> {
+    const patchPromise = new Promise<KeyedPatches>((resolve) =>
```

Comment:
I don't understand how this return value satisfies either part (keys or versions of those keys) of `keysToVersions`.